### PR TITLE
scalapb-runtime-grpc: optimize Marshaller for InProcessTransport

### PR DESCRIPTION
**Pull request by hugebdu**
resolves https://github.com/scalapb/ScalaPB/issues/1614

### CODE REVIEW

File: scalapb-runtime-grpc/src/main/scala/scalapb/grpc/ProtoInputStream.scala [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+import com.google.protobuf.CodedOutputStream
+import scalapb.GeneratedMessage
+
+import java.io.{ByteArrayInputStream, InputStream}
+
+/**
+  * Allows skipping serialization completely when the io.grpc.inprocess.InProcessTransport is used.
+  * Inspired by https://github.com/grpc/grpc-java/blob/master/protobuf-lite/src/main/java/io/grpc/protobuf/lite/ProtoInputStream.java
+  */
+class ProtoInputStream[T <: GeneratedMessage](msg: T, val marshaller: Marshaller[T]) extends InputStream {
```
Comment (thesamet) [inline scalapb-runtime-grpc/src/main/scala/scalapb/grpc/ProtoInputStream.scala:12]: Can you add unit tests for this class?

File: scalapb-runtime-grpc/src/main/scala/scalapb/grpc/Marshaller.scala [hunk 2]
```diff
@@ -6,10 +6,12 @@ import scalapb.{GeneratedMessage, GeneratedMessageCompanion, TypeMapper}
 
 class Marshaller[T <: GeneratedMessage](companion: GeneratedMessageCompanion[T])
     extends io.grpc.MethodDescriptor.Marshaller[T] {
-  override def stream(t: T): InputStream = new ByteArrayInputStream(t.toByteArray)
+  override def stream(t: T): InputStream = new ProtoInputStream[T](t, this)
 
-  override def parse(inputStream: InputStream): T =
-    companion.parseFrom(inputStream)
+  override def parse(inputStream: InputStream): T = inputStream match {
+    case pis: ProtoInputStream[T] if pis.marshaller == this => pis.message
```
Comment (thesamet) [inline scalapb-runtime-grpc/src/main/scala/scalapb/grpc/Marshaller.scala:12]: what's the reason for comparing the marshallers? if not needed, the `val marshaller` can be removed from the ProtoInputStream class.

File: project/Dependencies.scala [hunk 3]
```diff
@@ -60,6 +60,7 @@ object Dependencies {
   val grpcNetty     = "io.grpc" % "grpc-netty"           % versions.grpc
   val grpcServices  = "io.grpc" % "grpc-services"        % versions.grpc
   val grpcProtocGen = "io.grpc" % "protoc-gen-grpc-java" % versions.grpc
+  val grpcInprocess      = "io.grpc" % "grpc-inprocess"       % versions.grpc
```
[the remark this answers is not in the document]
Comment (hugebdu) [reply project/Dependencies.scala]: https://github.com/scalapb/ScalaPB/pull/1615/commits/dd570600e4a34cbefc6432fd175a1946bb6a8a70

File: scalapb-runtime-grpc/src/main/scala/scalapb/grpc/Marshaller.scala (hunk 2 above)
Comment (hugebdu) [reply scalapb-runtime-grpc/src/main/scala/scalapb/grpc/Marshaller.scala]: removed in https://github.com/scalapb/ScalaPB/pull/1615/commits/8d8256cffac1110abd44dcb75727c165c4d4ffbb

introduced it to reproduce the [following precaution logic](https://github.com/grpc/grpc-java/blob/master/protobuf-lite/src/main/java/io/grpc/protobuf/lite/ProtoLiteUtils.java#L175-L178), however seems to be redundant indeed.

File: scalapb-runtime-grpc/src/main/scala/scalapb/grpc/ProtoInputStream.scala (hunk 1 above)
Comment (hugebdu) [reply scalapb-runtime-grpc/src/main/scala/scalapb/grpc/ProtoInputStream.scala]: added in https://github.com/scalapb/ScalaPB/pull/1615/commits/eaec5ad36bf236c6993a999df6ab82e37948cebb

**Comment by hugebdu** [comment]
@thesamet formatted the code in https://github.com/scalapb/ScalaPB/pull/1615/commits/1aa83f125bba884a8613d46f4e91170e2c106ee6

thanks!

**Pull request merged**